x-pack/plugin/autoscaling/src/main/java/org/elasticsearch/xpack/autoscaling/action/ReservedAutoscalingPolicyAction.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.autoscaling.action;
+
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.reservedstate.ReservedClusterStateHandler;
+import org.elasticsearch.reservedstate.TransformState;
+import org.elasticsearch.xcontent.XContentParser;
+import org.elasticsearch.xcontent.XContentParserConfiguration;
+import org.elasticsearch.xpack.autoscaling.capacity.AutoscalingCalculateCapacityService;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.common.xcontent.XContentHelper.mapToXContentParser;
+
+/**
+ * This Action is the reserved state save version of RestPutAutoscalingPolicyHandler/RestDeleteAutoscalingPolicyHandler
+ * <p>
+ * It is used by the ReservedClusterStateService to add/update or remove autoscaling policies. Typical usage
+ * for this action is in the context of file based settings.
+ */
+public class ReservedAutoscalingPolicyAction implements ReservedClusterStateHandler<List<PutAutoscalingPolicyAction.Request>> {
+    public static final String NAME = "autoscaling";
+
+    private final AutoscalingCalculateCapacityService.Holder policyValidatorHolder;
+
+    /**
+     * Creates a ReservedAutoscalingPolicyAction
+     *
+     * @param policyValidatorHolder requires AutoscalingCalculateCapacityService.Holder for validation
+     */
+    public ReservedAutoscalingPolicyAction(final AutoscalingCalculateCapacityService.Holder policyValidatorHolder) {
+        this.policyValidatorHolder = policyValidatorHolder;
+    }
+
+    @Override
+    public String name() {
+        return NAME;
+    }
+
+    @SuppressWarnings("unchecked")
+    public Collection<PutAutoscalingPolicyAction.Request> prepare(Object input) {
+        List<PutAutoscalingPolicyAction.Request> policies = (List<PutAutoscalingPolicyAction.Request>) input;
+
+        for (var policy : policies) {
+            validate(policy);
+        }
+
+        return policies;
+    }
+
+    @Override
+    public TransformState transform(Object source, TransformState prevState) throws Exception {
```

Comment:
Could this not get the right type instead? Seems it is T at the interface level and the call in `ReservedStateUpdateTask` can be massaged to fix this. The service does "guarantee" it, since it only passes output of `fromXContent` to this method.

Reply:
On the caller side in `ReservedStateUpdateTask` I have a composite object containing the data for all of the handlers. I parse everything first in a map of String key to data type required by the handler, but I can't retain the type information because the data types are in classes provided by plugins and the server doesn't know of them.